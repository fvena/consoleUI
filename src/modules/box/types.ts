import type { Color } from "../style/types";

export type BorderStyle = "double" | "rounded" | "single";

export type TextAlign = "center" | "left" | "right";

export interface BoxOptions {
  align?: TextAlign;
  borderColor?: Color;
  borderStyle?: BorderStyle;
  horizontalPadding?: number;
  verticalPadding?: number;
  width?: number;
}

export interface BoxStyle {
  bottomLeft: string;
  bottomRight: string;
  horizontal: string;
  topLeft: string;
  topRight: string;
  vertical: string;
}
