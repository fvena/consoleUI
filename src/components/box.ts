import type {
  BorderStyle,
  BoxOptions,
  BoxStyle,
  Color,
  StripStylesFunction,
  StyleFunction,
} from "../core/types";
import { isBrowser } from "../utils/enviroment";
import { wrapText } from "../utils/wrap-text";
const BORDER_STYLES: Record<BorderStyle, BoxStyle> = {
  double: {
    bottomLeft: "╚",
    bottomRight: "╝",
    horizontal: "═",
    topLeft: "╔",
    topRight: "╗",
    vertical: "║",
  },
  rounded: {
    bottomLeft: "╰",
    bottomRight: "╯",
    horizontal: "─",
    topLeft: "╭",
    topRight: "╮",
    vertical: "│",
  },
  single: {
    bottomLeft: "└",
    bottomRight: "┘",
    horizontal: "─",
    topLeft: "┌",
    topRight: "┐",
    vertical: "│",
  },
};

const DEFAULT_OPTIONS: Required<BoxOptions> = {
  align: "left",
  borderColor: "gray",
  borderStyle: "rounded",
  horizontalPadding: 1,
  verticalPadding: 0,
  width: 80,
};

/**
 * Creates the box component factory with environment-specific styling
 * @param createStyle - Function to create environment-specific styles
 * @param stripStyles - Function to strip style characters
 * @returns Box component functions for the specific environment
 */
export function createBox(
  createStyle: (color: Color) => StyleFunction,
  stripStyles: StripStylesFunction,
) {
  /**
   * Creates a box with the specified content and options
   * @param content - The content to place inside the box
   * @param borderColor - The color of the border
   * @returns The styled box string
   */
  function box(content: string, borderColor?: Color): string {
    const boxOptions: Required<BoxOptions> = {
      ...DEFAULT_OPTIONS,
      ...(borderColor ? { borderColor } : {}),
    };

    return makeBox(boxOptions)(content);
  }

  /**
   * Creates a reusable box style function
   * @param options - The box options
   * @returns A function that creates a box with the specified style
   */
  function makeBox(options: BoxOptions): StyleFunction {
    const boxOptions: Required<BoxOptions> = { ...DEFAULT_OPTIONS, ...options };
    const borderStyle = BORDER_STYLES[boxOptions.borderStyle];
    const horizontalPadding = " ".repeat(boxOptions.horizontalPadding);
    const borderColor = createStyle(boxOptions.borderColor);
    const contentWidth = boxOptions.width - boxOptions.horizontalPadding * 2 - 2; // Account for borders and horizontal padding

    return (content: string): string => {
      // Split content into lines and wrap each line to contentWidth
      const inputLines = content.split("\n");
      const lines: string[] = [];
      for (const line of inputLines) {
        lines.push(...wrapText(line, contentWidth, stripStyles));
      }

      const boxWidth = boxOptions.width;
      const horizontalLine = borderStyle.horizontal.repeat(boxWidth - 2);

      const top = borderColor(borderStyle.topLeft + horizontalLine + borderStyle.topRight);
      const bottom = borderColor(borderStyle.bottomLeft + horizontalLine + borderStyle.bottomRight);

      const paddedLines = lines.map((line) => {
        const strippedLine = stripStyles(line);
        const remainingSpace = contentWidth - strippedLine.length;

        let alignedLine = line;
        if (boxOptions.align === "center" && remainingSpace > 0) {
          const leftPadding = Math.floor(remainingSpace / 2);
          const rightPadding = remainingSpace - leftPadding;
          alignedLine = " ".repeat(leftPadding) + line + " ".repeat(rightPadding);
        } else if (boxOptions.align === "right" && remainingSpace > 0) {
          alignedLine = " ".repeat(remainingSpace) + line;
        } else {
          // Left alignment (default)
          alignedLine = line + " ".repeat(remainingSpace);
        }

        return (
          borderColor(borderStyle.vertical) +
          horizontalPadding +
          alignedLine +
          horizontalPadding +
          borderColor(borderStyle.vertical)
        );
      });

      // Add vertical padding
      const emptyLine =
        borderColor(borderStyle.vertical) +
        " ".repeat(boxWidth - 2) +
        borderColor(borderStyle.vertical);
      const verticalPadding = Array.from({ length: boxOptions.verticalPadding }, () => emptyLine);

      const allLines = [top, ...verticalPadding, ...paddedLines, ...verticalPadding, bottom];

      // Wrap the entire box with monospace font style
      if (isBrowser()) {
        const monospaceStart = "__STYLE_MONOSPACE__";
        const monospaceEnd = "__STYLE_RESET__";
        return monospaceStart + allLines.join("\n") + monospaceEnd;
      }
      return allLines.join("\n");
    };
  }

  return { box, makeBox };
}
