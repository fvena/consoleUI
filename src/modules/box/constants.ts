import type { BorderStyle, BoxOptions, BoxStyle } from "./types";
/**
 * The border styles for the box component
 */
export const BORDER_STYLES: Record<BorderStyle, BoxStyle> = {
  double: {
    bottomLeft: "╚",
    bottomRight: "╝",
    horizontal: "═",
    topLeft: "╔",
    topRight: "╗",
    vertical: "║",
  },
  rounded: {
    bottomLeft: "╰",
    bottomRight: "╯",
    horizontal: "─",
    topLeft: "╭",
    topRight: "╮",
    vertical: "│",
  },
  single: {
    bottomLeft: "└",
    bottomRight: "┘",
    horizontal: "─",
    topLeft: "┌",
    topRight: "┐",
    vertical: "│",
  },
};

/**
 * The default options for the box component
 */
export const DEFAULT_OPTIONS: Required<BoxOptions> = {
  align: "left",
  borderColor: "gray",
  borderStyle: "rounded",
  horizontalPadding: 1,
  verticalPadding: 0,
  width: 80,
};
