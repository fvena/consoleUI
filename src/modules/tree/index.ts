import type { Color, StyleFunction } from "../style/type";
import type { TreeNode, TreeOptions } from "./types";
import { isBrowser } from "../../utils/enviroment";
import { DEFAULT_OPTIONS, DEFAULT_SYMBOLS } from "./constants";

/**
 * Creates the tree component factory with environment-specific styling
 * @param createStyle - Function to create environment-specific styles
 * @returns Tree component functions for the specified environment
 */
export function createTree(createStyle: (color: Color) => StyleFunction) {
  /**
   * Creates a tree visualization with the specified content and options
   * @param content - The tree content to visualize (string or TreeNode)
   * @param color - The color of the tree lines
   * @returns The styled tree string
   */
  function tree(content: TreeNode | TreeNode[], showIcons = true, color?: Color): string {
    const treeOptions: Required<TreeOptions> = {
      ...DEFAULT_OPTIONS,
      showIcons,
      ...(color ? { color } : {}),
    };

    return makeTree(treeOptions)(content);
  }

  /**
   * Creates a reusable tree style function
   * @param options - The tree options
   * @returns A function that creates a tree with the specified style
   */
  function makeTree(options: TreeOptions): (content: TreeNode | TreeNode[]) => string {
    const treeOptions: Required<TreeOptions> = { ...DEFAULT_OPTIONS, ...options };

    return (content: TreeNode | TreeNode[]): string => {
      let result = "";

      if (Array.isArray(content)) {
        // Handle array of nodes
        for (const [index, node] of content.entries()) {
          const isLast = index === content.length - 1;
          result += renderNode(node, treeOptions, 0, isLast, "");
        }
      } else {
        // Handle single node
        result = renderNode(content, treeOptions, 0, true, "");
      }

      return result;
    };
  }

  /**
   * Renders a tree node with the specified depth
   * @param node - The tree node to render
   * @param treeOptions - The tree options
   * @param depth - The current depth of the node
   * @param isLast - Whether this is the last child node
   * @param prefix - The prefix to add before this node
   * @returns The styled tree node string
   */
  function renderNode(
    node: TreeNode,
    treeOptions: Required<TreeOptions>,
    depth = 0,
    isLast = true,
    prefix = "",
  ): string {
    const isRoot = depth === 0;
    const hasChildren = node.children && node.children.length > 0;

    // Style the node symbol
    const treeColor = createStyle(treeOptions.color);
    let nodeSymbol = "";
    if (!isRoot) {
      nodeSymbol = isLast ? DEFAULT_SYMBOLS.corner : DEFAULT_SYMBOLS.branch;
    }
    let styledNodeSymbol = "";

    // Wrap the node symbol with monospace font style in browser environments,
    // because the font style is not monospace by default in the terminal,
    // and the tree is not aligned properly.
    if (isBrowser()) {
      const monospaceStart = "__STYLE_MONOSPACE__";
      const monospaceEnd = "__STYLE_RESET__";
      styledNodeSymbol = monospaceStart + prefix + treeColor(nodeSymbol) + monospaceEnd;
    } else {
      styledNodeSymbol = prefix + treeColor(nodeSymbol);
    }

    // Use custom icon if provided, otherwise use default icons based on node type
    let icon = "";
    if (treeOptions.showIcons) {
      icon = node.icon ?? (hasChildren ? DEFAULT_SYMBOLS.folderIcon : DEFAULT_SYMBOLS.fileIcon);
      icon += " "; // Add a space after the icon
    }

    // Line for the current node
    let result = styledNodeSymbol + icon + node.label + "\n";

    // Add vertical lines for children nodes if any
    if (hasChildren && node.children) {
      // Adjust child prefix based on whether this is the root node
      const childPrefix = isRoot
        ? ""
        : prefix + (isLast ? "    " : treeColor(DEFAULT_SYMBOLS.vertical));

      for (const [index, child] of node.children.entries()) {
        const isLastChild = index === node.children.length - 1;
        result += renderNode(child, treeOptions, depth + 1, isLastChild, childPrefix);
      }
    }

    return result;
  }

  return { makeTree, tree };
}
