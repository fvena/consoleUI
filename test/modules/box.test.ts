import type { Color } from "../../src/modules/style/type";
import { describe, expect, it } from "vitest";
import { createBox } from "../../src/modules/box";

/**
 * Creates a mock style function for testing that ignores color and returns text unchanged
 * @param _color - The color parameter (intentionally unused in tests)
 * @returns A mock style function
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars -- Parameter is required by type but unused in mock
function createMockStyle(_color: Color) {
  return function mockStyle(text: string): string {
    // We ignore the color parameter in tests as we're just testing the box structure
    return text;
  };
}

const { box, makeBox } = createBox(createMockStyle, () => "");

describe("Box Component", () => {
  describe("box", () => {
    it("should create a box with default options", () => {
      const result = box("Hello");
      expect(result).toMatchSnapshot();
    });

    it("should create a box with custom border color", () => {
      const result = box("Hello", "blue");
      expect(result).toMatchSnapshot();
    });

    it("should create a box with custom options", () => {
      const result = box("Hello", "red");
      expect(result).toMatchSnapshot();
    });

    it("should handle multi-line content", () => {
      const result = box("Line 1\nLine 2\nLine 3");
      expect(result).toMatchSnapshot();
    });

    it("should handle empty content", () => {
      const result = box("");
      expect(result).toMatchSnapshot();
    });

    it("should wrap text at default width", () => {
      const longText =
        "This is a very long text that should be wrapped automatically at the default width of 80 characters because it exceeds that limit significantly";
      const result = box(longText);
      expect(result).toMatchSnapshot();
    });
  });

  describe("makeBox", () => {
    it("should create a reusable box style", () => {
      const customBox = makeBox({
        borderColor: "green",
        borderStyle: "rounded",
        horizontalPadding: 0,
        verticalPadding: 0,
      });

      const result = customBox("Hello");
      expect(result).toMatchSnapshot();
    });

    it("should handle different content with same style", () => {
      const customBox = makeBox({
        borderColor: "yellow",
        borderStyle: "single",
        horizontalPadding: 1,
        verticalPadding: 1,
      });

      const result1 = customBox("First");
      const result2 = customBox("Second");

      expect(result1).toMatchSnapshot("First content");
      expect(result2).toMatchSnapshot("Second content");
    });

    it("should handle multi-line content with custom style", () => {
      const customBox = makeBox({
        borderColor: "magenta",
        borderStyle: "double",
        horizontalPadding: 2,
        verticalPadding: 2,
      });

      const result = customBox("Line 1\nLine 2\nLine 3");
      expect(result).toMatchSnapshot();
    });

    it("should respect custom width", () => {
      const customBox = makeBox({
        borderColor: "cyan",
        width: 40,
      });

      const longText =
        "This text should be wrapped at 40 characters width instead of the default 80";
      const result = customBox(longText);
      expect(result).toMatchSnapshot();
    });

    it("should handle very long words", () => {
      const customBox = makeBox({
        width: 20,
      });

      const textWithLongWord = "Normal ThisIsAVeryLongWordThatShouldBeSplitAcrossMultipleLines end";
      const result = customBox(textWithLongWord);
      expect(result).toMatchSnapshot();
    });

    it("should handle multiple paragraphs with wrapping", () => {
      const customBox = makeBox({
        width: 30,
      });

      const multiParagraph =
        "First paragraph that should wrap.\n\nSecond paragraph that should also wrap across multiple lines.";
      const result = customBox(multiParagraph);
      expect(result).toMatchSnapshot();
    });
  });
});
