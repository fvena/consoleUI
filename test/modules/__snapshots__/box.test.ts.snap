// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Box Component > box > should create a box with custom border color 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────╮
│ Hello                                                                             │
╰──────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`Box Component > box > should create a box with custom options 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────╮
│ Hello                                                                             │
╰──────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`Box Component > box > should create a box with default options 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────╮
│ Hello                                                                             │
╰──────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`Box Component > box > should handle empty content 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────╮
╰──────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`Box Component > box > should handle multi-line content 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────╮
│ Line 1                                                                             │
│ Line 2                                                                             │
│ Line 3                                                                             │
╰──────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`Box Component > box > should wrap text at default width 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────╮
│ This is a very long text that should be wrapped automatically at the default width of 80 characters because it exceeds that limit significantly                                                                             │
╰──────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`Box Component > makeBox > should create a reusable box style 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────╮
│Hello                                                                              │
╰──────────────────────────────────────────────────────────────────────────────╯"
`;

exports[`Box Component > makeBox > should handle different content with same style > First content 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────┐
│                                                                              │
│ First                                                                             │
│                                                                              │
└──────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`Box Component > makeBox > should handle different content with same style > Second content 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────┐
│                                                                              │
│ Second                                                                             │
│                                                                              │
└──────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`Box Component > makeBox > should handle multi-line content with custom style 1`] = `
"╔══════════════════════════════════════════════════════════════════════════════╗
║                                                                              ║
║                                                                              ║
║  Line 1                                                                            ║
║  Line 2                                                                            ║
║  Line 3                                                                            ║
║                                                                              ║
║                                                                              ║
╚══════════════════════════════════════════════════════════════════════════════╝"
`;

exports[`Box Component > makeBox > should handle multiple paragraphs with wrapping 1`] = `
"╭────────────────────────────╮
│ First paragraph that should wrap.                           │
│ Second paragraph that should also wrap across multiple lines.                           │
╰────────────────────────────╯"
`;

exports[`Box Component > makeBox > should handle very long words 1`] = `
"╭──────────────────╮
│ Normal ThisIsAVeryLongWordThatShouldBeSplitAcrossMultipleLines end                 │
╰──────────────────╯"
`;

exports[`Box Component > makeBox > should respect custom width 1`] = `
"╭──────────────────────────────────────╮
│ This text should be wrapped at 40 characters width instead of the default 80                                     │
╰──────────────────────────────────────╯"
`;
